### PR TITLE
tests, overlord: extend unit tests, extend spread tests to cover full command line support

### PR DESCRIPTION
**Pull request by bboozzoo**
~~Stacked on top of #10148 and #10156.~~

The branch brings together everything for cmdline.full and cmdline.extra suppot, extends the unit tests and improves spread tests to execute full command line scenarios.

### CODE REVIEW

File: tests/nested/manual/core20-boot-config-update/task.yaml [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+  nested_exec "cat /proc/cmdline" > system.cmdline
+
+  case "$VARIANT" in
+    no-gadget)
+      echo "check that new static command line element is present, but nothing comes from the gadget"
+      MATCH bootassetstesting < system.cmdline
+      not MATCH 'hello from test' < system.cmdline
+      ;;
+    gadget-extra)
+      echo "check that new static command line element is present, extra bits from the gadget"
```
Comment (anonymouse64) [inline tests/nested/manual/core20-boot-config-update/task.yaml:123]: very nice

**Review by anonymouse64 — COMMENTED** [review]
Last 3 commits look good to me, thanks for the spread tests

File: boot/boot_test.go [hunk 2]
```diff
@@ -3749,3 +3764,138 @@ func (s *bootKernelCommandLineSuite) TestCommandLineUpdateUC20TransitionFullExtr
...
+	c.Check(s.resealCalls, Equals, 1)
+	c.Check(s.bootloader.SetBootVarsCalls, Equals, 1)
+	// all done, modeenv
+	m, err = boot.ReadModeenv("")
+	c.Assert(err, IsNil)
+	c.Check(m.CurrentKernelCommandLines, DeepEquals, boot.BootCommandLines{
+		"snapd_recovery_mode=run static mocked panic=-1",
+		"snapd_recovery_mode=run static mocked panic=-1 extra args",
+	})
+	args, err := s.bootloader.GetBootVars("snapd_extra_cmdline_args", "snapd_full_cmdline_args")
```
Comment (mvo5) [inline boot/boot_test.go:3894]: Should we check earlier that the canary is still intact in this test?

File: tests/nested/manual/core20-boot-config-update/task.yaml [hunk 3]
```diff
@@ -0,0 +1,155 @@
+summary: Check that the boot config is correctly updated when snapd is refreshed
```
Comment (pedronis) [inline tests/nested/manual/core20-boot-config-update/task.yaml:1]: why is it moved to manual?

File: boot/boot_test.go [hunk 4]
```diff
@@ -3749,3 +3764,138 @@ func (s *bootKernelCommandLineSuite) TestCommandLineUpdateUC20TransitionFullExtr
...
+	err = boot.MarkBootSuccessful(s.uc20dev)
+	c.Assert(err, IsNil)
+	// we resealed after reboot again
+	c.Check(s.resealCalls, Equals, 2)
+	m, err = boot.ReadModeenv("")
+	c.Assert(err, IsNil)
+	c.Check(m.CurrentKernelCommandLines, DeepEquals, boot.BootCommandLines{
+		"snapd_recovery_mode=run static mocked panic=-1",
+	})
+
```
Comment (pedronis) [inline boot/boot_test.go:3877]: we should also check what happens if we reboot after setting the vars but before setting the task to done

**Review by pedronis — COMMENTED** [review]
thanks for this, some comments and questions

File: tests/nested/manual/core20-boot-config-update/task.yaml (hunk 3 above)
Comment (bboozzoo) [reply tests/nested/manual/core20-boot-config-update/task.yaml:1]: The image is built with customized gadget, whereas the nested core suite tests execute with a standard pc snap that that only got the shim signed by the test setup.

File: tests/nested/manual/core20-boot-config-update/task.yaml [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+  esac
+
+  echo "Install new (unasserted) snapd and wait for reboot/change finishing"
+  boot_id="$( nested_get_boot_id )"
+  REMOTE_CHG_ID=$(nested_exec sudo snap install --dangerous snapd-boot-config-update.snap --no-wait)
+  nested_exec sudo snap watch "${REMOTE_CHG_ID}"
+  if [ "$VARIANT" != "gadget-full" ]; then
+    # reboot is automatically requested by snapd only if part of the command
+    # line is changed, in case of gadget overriding the command line fully
+    # (gadget-full variant), there is no resealing and no reboot
```
Comment (mvo5) [inline tests/nested/manual/core20-boot-config-update/task.yaml:106]: Maybe a small comment here *why* no resealing and no reboot?

**Review by mvo5 — APPROVED** [review]
Looks good, thank you! Some questions inline

File: boot/boot_test.go (hunk 2 above)
Comment (bboozzoo) [reply boot/boot_test.go:3894]: Earlier bits either panic'ed before calling SetBootVars or right when entering it (due to SetMockPanic), in either case the test checks that SetBootVarsCalls is 0.

File: tests/nested/manual/core20-boot-config-update/task.yaml (hunk 5 above)
Comment (bboozzoo) [reply tests/nested/manual/core20-boot-config-update/task.yaml:106]: Tweaked the comment a bit.

File: boot/boot_test.go [hunk 6]
```diff
@@ -3749,3 +3764,138 @@ func (s *bootKernelCommandLineSuite) TestCommandLineUpdateUC20TransitionFullExtr
...
+	err = s.bootloader.SetBootVars(map[string]string{
+		// those are intentionally filled by the test
+		"snapd_extra_cmdline_args": "canary",
+		"snapd_full_cmdline_args":  "canary",
+	})
+	c.Assert(err, IsNil)
+	s.bootloader.SetBootVarsCalls = 0
+
+	restoreBootloaderNoPanic := s.bootloader.SetMockPanic("SetBootVars")
```
[the remark this answers is not in the document]
Comment (bboozzoo) [reply boot/boot_test.go:3803]: added

File: bootloader/bootloadertest/bootloadertest.go [hunk 7]
```diff
@@ -148,6 +148,29 @@ func (b *MockBootloader) InstallBootConfig(gadgetDir string, opts *bootloader.Op
...
+	case "SetBootVars", "GetBootVars",
+		"EnableKernel", "EnableTryKernel", "Kernel", "TryKernel", "DisableTryKernel":
+
+		old := b.panicMethods[f]
+		b.panicMethods[f] = true
+		return func() {
+			b.panicMethods[f] = old
+		}
+	default:
+		panic(fmt.Sprintf("unknown ExtractedRunKernelImageBootloader method %q to mock reboot via panic for", f))
```
[the remark this answers is not in the document]
Comment (bboozzoo) [reply bootloader/bootloadertest/bootloadertest.go:170]: tweaked

File: bootloader/bootloadertest/bootloadertest.go [hunk 8]
```diff
@@ -148,6 +148,29 @@ func (b *MockBootloader) InstallBootConfig(gadgetDir string, opts *bootloader.Op
 	return b.InstallBootConfigErr
 }
 
+// SetMockPanic allows setting any method in the Bootloader interface or derived
+// interface to panic instead of returning. This allows one to test what would
+// happen if the system was rebooted during execution of a particular function.
+// Specifically, the panic will be done immediately entering the function so
+// setting SetBootVars to panic will emulate a reboot before any boot vars are
+// set persistently
+func (b *MockBootloader) SetMockPanic(f string) (restore func()) {
```
[the remark this answers is not in the document]
Comment (bboozzoo) [reply bootloader/bootloadertest/bootloadertest.go:157]: renamed

File: boot/boot_test.go (hunk 4 above)
Comment (bboozzoo) [reply boot/boot_test.go:3878]: Updated. I've tried to add a devicestate level test, but could not come up with reasonable way to handle a panic in task runner. Instead I've added one more boot package test that panics after setting boot vars, which should mimic the scenario in which we reboot before marking the task as done.

**Review by pedronis — APPROVED** [review]
thanks

**Comment by mvo5** [comment]
Merging, errors are unrelated.

**Pull request merged**